docs_src/design/adr/0023-North-South-Messaging.md

```diff
@@ -0,0 +1,271 @@
+
+Today, communications from a 3rd party system (enterprise application, cloud application, etc.) to EdgeX in order to acuate a device or get the latest information from a sensor is accomplished via REST.  The 3rd party system makes a REST call of the command service which then relays a request to a device service also using REST.  There is no built in means to make a message-based request of EdgeX or the devices/sensors it manages.
+
+In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.
+
+### Core Command as Message Bus Bridge
+
+The core command service will serve as the EdgeX entry point for external, north-to-south message bus requests to the south side.
+
+![image](command-msg.png)
+
+3rd party systems should not be granted access to the EdgeX internal message bus.  Therefore, in order to implement north to south communications via message bus, the command service needs to take messages from the 3rd party or external message bus and pass them internally onto the EdgeX internal message bus where they can eventually be routed to the device services and then on to the devices/sensors (southside).
+
+In reverse, response messages from the southside will also be sent through the internal EdgeX message bus to the command service where they can then be bridged to the external message bus and respond to the 3rd party system requester.
+
+!!! Note
+    If command did not serve as this external to internal (and vice versa) bridge, then each device service would need to be able to serve as their own external message bus to internal message bus bridge which is inefficient.  Also note that eKuiper is allowed access directly to the internal EdgeX message bus.  This is a special circumstance of 3rd party external system communication as eKuiper is a sister project that is deemed the EdgeX reference implementation rules engine.  In future releases of EdgeX, even eKuiper may be routed through an external to internal message bus bridge for better decoupling and security.
+
+Any protocol translation or relay service (REST-MQTT, MQTT-REST, REST-REST, MQTT-MQTT) should pass along authentication tokens.
```

Comment:
Actually, the Kong Auth token does flow through.  MQTT credentials are not part of the MQTT payload.  The intent of this requirement is that we don't want to change the security context of the original request.  When microservices (eventually) start to authenticate requests, we don't want them to reject messages through the bridge because it stripped information about the original caller.  Similarly, we don't want to give the bridge a super-admin token because then the bridge would become a privilege escalation mechanism.  Also, it doesn't seem like a good idea to make the bridge responsible for 1:1 mappings from external credential into internal credentials, as that would require the bridge to be able to fabricate credentials.  That leaves passthrough as the most reasonable choice.

Reply:
@bnevis-i , still doesn't make sense to me for the External MQTT -> Internal MessageBus or HTTP -> Internal MessageBus. We should discuss on next security WG meeting.

Reply:
As discusssed in security WG meeting on 2/2/2022, drop this line for this ADR.  Microservice authentication ADR will be extended to address message-bus based communication in addition to REST.   The current situation only applies to core-command listening to an MQTT bus directly as an alternative interface.  Current proposal is not to have a microservice that generically bridges message busses.